File: priv/combined_metrics/samples/variable_naming/no_hungarian_notation/good/inventory.js
// Inventory management without Hungarian notation prefixes.
// GOOD: name, isActive, count, items, age, user, callback, price — no type prefixes.

function addItem(item) {
  const name = item.name;
  const sku = item.sku;
  const quantity = item.quantity;
  const price = item.price;
  const isActive = item.active ?? true;

  if (!name || name.trim() === '') {
    return { ok: false, error: 'Name is required' };
  }

  return {
    ok: true,
    data: { id: generateId(), name, sku, quantity, price, active: isActive },
  };
}

function updateStock(itemId, delta) {
  const item = fetchItem(itemId);
  const newQuantity = item.quantity + delta;

  if (newQuantity < 0) {
    return { ok: false, error: 'Insufficient stock' };
  }

  return { ok: true, data: { ...item, quantity: newQuantity } };
}

function searchItems(query, items) {
  const lowerQuery = query.toLowerCase();

  return items.filter(item =>
    item.name.toLowerCase().includes(lowerQuery) ||
    item.sku.toLowerCase().includes(lowerQuery)
  );
}

function calculateValue(items) {
  return items.reduce((acc, item) => acc + item.quantity * item.price, 0);
}

function applyDiscount(items, discountRate, filter) {
  return items
    .filter(filter)
    .map(item => {
      const newPrice = Math.round(item.price * (1 - discountRate) * 100) / 100;
      return { ...item, price: newPrice };
    });
}

function groupByCategory(items) {
  return items.reduce((acc, item) => {
    const category = item.category;
    acc[category] = acc[category] || [];
    acc[category].push(item);
    return acc;
  }, {});
}

function lowStockReport(items, threshold) {
  const includeInactive = false;

  return items
    .filter(item => item.quantity <= threshold && (includeInactive || item.active))
    .sort((a, b) => a.quantity - b.quantity);
}

function importItems(rawData, transform, validate) {
  const transformed = rawData.map(transform);
  const valid = transformed.filter(validate);
  const imported = valid.length;
  const skipped = rawData.length - imported;

  return { imported, skipped, items: valid };
}

function fetchItem(id) { return { id, name: 'Item', sku: 'SKU', quantity: 10, price: 9.99, active: true, category: 'misc' }; }
function generateId() { return Math.random().toString(36).slice(2); }

module.exports = { addItem, updateStock, searchItems, calculateValue, applyDiscount, groupByCategory, lowStockReport, importItems };

File: priv/combined_metrics/samples/variable_naming/no_hungarian_notation/good/inventory.rb
# Inventory management without Hungarian notation prefixes.
# GOOD: name, is_active, count, items, age, user, callback, price — no type prefixes.

class InventoryGood
  def add_item(item)
    name = item[:name]
    sku = item[:sku]
    quantity = item[:quantity]
    price = item[:price]
    is_active = item.fetch(:active, true)

    return { ok: false, error: 'Name is required' } if name.nil? || name.strip.empty?

    {
      ok: true,
      data: { id: generate_id, name: name, sku: sku, quantity: quantity, price: price, active: is_active }
    }
  end

  def update_stock(item_id, delta)
    item = fetch_item(item_id)
    new_quantity = item[:quantity] + delta

    return { ok: false, error: 'Insufficient stock' } if new_quantity < 0

    { ok: true, data: item.merge(quantity: new_quantity) }
  end

  def search_items(query, items)
    lower_query = query.downcase

    items.select do |item|
      item[:name].downcase.include?(lower_query) ||
        item[:sku].downcase.include?(lower_query)
    end
  end

  def calculate_value(items)
    items.sum { |item| item[:quantity] * item[:price] }
  end

  def apply_discount(items, discount_rate, filter)
    items
      .select(&filter)
      .map do |item|
        new_price = (item[:price] * (1 - discount_rate)).round(2)
        item.merge(price: new_price)
      end
  end

  def group_by_category(items)
    items.group_by { |item| item[:category] }
  end

  def low_stock_report(items, threshold)
    include_inactive = false

    items
      .select { |item| item[:quantity] <= threshold && (include_inactive || item[:active]) }
      .sort_by { |item| item[:quantity] }
  end

  def import_items(raw_data, transform, validate)
    transformed = raw_data.map(&transform)
    valid = transformed.select(&validate)
    imported = valid.length
    skipped = raw_data.length - imported

    { imported: imported, skipped: skipped, items: valid }
  end

  private

  def fetch_item(id)
    { id: id, name: 'Item', sku: 'SKU', quantity: 10, price: 9.99, active: true, category: 'misc' }
  end

  def generate_id = SecureRandom.hex(8)
end
